Add support for LTC4162L

mrcsosa opened the pull request:
## Pull Request Description

Please replace this with a detailed description and motivation of the changes. 
You can tick the checkboxes below with an 'x' between square brackets or just check them after publishing the PR. 
If this PR contains a breaking change, list dependent PRs and try to push all related PRs at the same time.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

CiprianRegus commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+ * @param val - The value to write to the register.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ltc4162l_reg_write(struct ltc4162l_desc * dev, uint8_t addr, uint16_t val)
```
If only the lower byte is used, does `val` need to be 16 bits wide?

mrcsosa replied:
Hi, I updated the write function to write 16-bits wide instead of 8

CiprianRegus commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+ */
+int ltc4162l_reg_write(struct ltc4162l_desc * dev, uint8_t addr, uint16_t val)
+{
+	uint8_t buff [] = {addr, val & 0xFF};
```
Remove the space between `buff` and `[]`.

CiprianRegus commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+int ltc4162l_reg_read(struct ltc4162l_desc * dev, uint8_t addr, uint16_t *val)
+{
+	int ret;
+	uint8_t uval [2];
```
Remove the space

CiprianRegus commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+	if (ret)
+		return ret;
+
+	switch (dev->type) {
+	case ID_LTC4162L:
+	case ID_LTC4162F:
+		regval *= 1924;
+		regval *= cell_count;
+		regval /= 10;
+		*val = regval;
+		return 0;
+	case ID_LTC4162S:
+		regval *= 3848;
+		regval *= cell_count / 2;
+		regval /= 10;
+		*val = regval;
+		return 0;
+	case ID_LTC4015:
+		regval *= chem_type <= 6 ? 192264 : 128176;
+		regval *= cell_count;
+		regval /= 1000;
+		*val = regval;
+		return 0;
+	default:
+		return -EINVAL;
+	}
```
Add some macros for these numbers (and maybe a comment describing what are the units).

mrcsosa replied:
I have created new fcompute unctions to support chip_info properties and added some description for the units

CiprianRegus commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+	if (ret)
+		return ret;
+
+	ret = (int16_t)(regval & 0xFFFF);
```
I'd say `ret = no_os_sign_extend32(regval, 15);` better explains what you're trying to do.

kister-jimenez commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+{
+	uint16_t regval;
+
+	regval = LTC4162L_MULT_FRAC(val, dev->rsnsi, 50000u);
+	regval /= 10000u;
```
Put chip dependent constants in chip_info especially if these values are different between variants.

kister-jimenez commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+	switch (dev->type) {
+	case ID_LTC4162L:
+	case ID_LTC4162F:
+		regval *= 1924;
+		regval *= cell_count;
+		regval /= 10;
+		*val = regval;
+		return 0;
+	case ID_LTC4162S:
+		regval *= 3848;
+		regval *= cell_count / 2;
+		regval /= 10;
+		*val = regval;
+		return 0;
+	case ID_LTC4015:
+		regval *= chem_type <= 6 ? 192264 : 128176;
+		regval *= cell_count;
+		regval /= 1000;
+		*val = regval;
```
Can you check if you can use chip_info struct to house chip_dependent constants?

kister-jimenez commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+
+	ret = (int16_t)(regval & 0xFFFF);
+
+	*val = 100 * LTC4162L_MULT_FRAC(ret, dev->type == ID_LTC4015 ? 146487 : 14660,
```
check if you can use a chip_info variable to store chip-dependent constants. This way you don't have to have an inline  if/else.

mrcsosa replied:
moved to chip_info function and created new computation function

kister-jimenez commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+		return ret;
+
+	ret = (int16_t)(regval & 0xFFFF);
+	ret *= dev->type == ID_LTC4015 ? 146487 : 14660;
```
ditto

mrcsosa replied:
moved to chip_info function and created new computation function

kister-jimenez commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+	ret = (int16_t)(regval & 0xFFFF);
+	ret *= dev->type == ID_LTC4015 ? 146487 : 14660;
+	ret /= dev->rsnsi;
+	ret *= dev->type == ID_LTC4015 ? 1000 : 100;
```
ditto

mrcsosa replied:
moved to chip_info function and created new computation function

kister-jimenez commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+
+	switch (dev->type) {
+	case ID_LTC4162L:
+		voltage = 3812500 + (regval * 12500);
+		voltage *= cell_count;
+		*val = voltage;
+
+		return 0;
+	case ID_LTC4162F:
+		voltage = 3412500 + (regval * 12500);
+		voltage *= cell_count;
+		*val = voltage;
+
+		return 0;
+	case ID_LTC4162S:
+		voltage = 6000000 + (regval * 28571);
+		voltage *= cell_count/2;
+		*val = voltage;
+
+		return 0;
+	case ID_LTC4015:
+		if (chem_type <= 3)
+			voltage = 3812500 + (regval * 12500);
+		else if (chem_type > 3 || chem_type < 7)
+			voltage = 3412500 + (regval * 12500);
+		else
+			voltage = 2000000 + (regval * 9523);
```
ditto

mrcsosa replied:
moved to chip_info function and created new computation function

kister-jimenez commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+
+	switch (dev->type) {
+	case ID_LTC4162L:
+		base_voltage = 3812500;
+		scale_factor = 12500;
+		range = 31;
+		break;
+	case ID_LTC4162F:
+		base_voltage = 3412500;
+		scale_factor = 12500;
+		range = 31;
+		break;
+	case ID_LTC4162S:
+		base_voltage = 6000000;
+		scale_factor = 28571;
+		range = 31;
+		break;
+	case ID_LTC4015:
+		if (chem_type <= 3) {
+			base_voltage = 3812500;
+			scale_factor = 12500;
+			range = 31;
+		} else if (chem_type < 7) {
+			base_voltage = 3412500;
+			scale_factor = 12500;
+			range = 31;
+		} else {
+			base_voltage = 2000000;
+			scale_factor = 9523;
+			range = 35;
```
ditto

mrcsosa replied:
moved to chip_info function and created new computation function

mrcsosa commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,829 @@
+					   LTC4162L_EN_C_OVER_X_TERM_BIT, 0);
+	}
+
+	regval = LTC4162L_MULT_FRAC(val, dev->rsnsb, 14660u);
+	regval /= 100u;
```
moved to chip_info function and created new computation function

buha commented:
Please fetch the newest main on your local machine and then `git rebase main` from this branch, it might fix some CI failures that are unrelated to your PR.

mrcsosa commented:
Changelogs:

- Modified the write function to write 16-bits wide instead of 8 (pushed an older version before)
- Created chip_info and modified functions with computation
- Created computation functions for more scalability
- Added computation macros and descriptions
- Fixed Astyle issues

kister-jimenez commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,860 @@
+				enum ltc4162l_command cmd)
+{
+	switch (cmd) {
+	case LTC4162L_CMD_GET_VBAT: /* x = regval, y = cell count, z = chem type*/
+		return (x * (z <= 6 ? 192264 : 128176) * y) / 1000;
+	case LTC4162L_CMD_GET_VCHARGE:
+		return ((z <= 3 ? (3812500 + x * 12500) : (z < 7 ? (3412500 + x * 12500) :
+				(2000000 + x * 9523))) * y);
+	case LTC4162L_CMD_GET_DIE_TEMP:
+		return x - 12010 * 1000 / 456;
```
ditto

kister-jimenez commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,860 @@
+static uint16_t ltc4162s_compute(uint16_t x, uint16_t y, uint16_t z,
+				 enum ltc4162l_command cmd)
+{
+	switch (cmd) {
+	case LTC4162L_CMD_GET_VBAT: /* x = regval, y = cell count*/
+		return x * y / 2 * 3848 / 10;
+	case LTC4162L_CMD_GET_VCHARGE:
+		return (6000000 + (x * 28571)) * y / 2;
+	case LTC4162L_CMD_GET_DIE_TEMP:
+		return x * 215 / 100 - 26440;
+	default:
```
ditto

kister-jimenez commented on drivers/power/ltc4162l/ltc4162l.c:
```diff
@@ -0,0 +1,860 @@
+	if (ret)
+		return ret;
+
+	if (dev->id == ID_LTC4015) {
+		*val = (no_os_field_get(LTC4162L_RUN_BSR_BIT, regval)) ? 1 : 0;
+
+	} else {
+		*val = (no_os_field_get(LTC4162L_FORCE_TELEMETRY_BIT, regval)) ? 1 : 0;
+	}
```
no need for brackets since if and else both have single statements.

mrcsosa commented:
V3

Overhauled the computation functions and chip_info:
> instead of using general compute function -> created separate functions for each of the computations.
> directly uses the chip_info parameters instead of hardcoding the compute values.

kister-jimenez commented:
LGTM. Xilinx issue in the CI is not related.

buha commented on drivers/power/ltc4162l/ltc4162l.h:
```diff
@@ -0,0 +1,266 @@
+#define LTC4015_VCHARGE_RANGE			34125
+
+
+#define LTC4162L_MULT_FRAC(x, mult, div) ((x * mult) / div)
```
i don't see the need of this macro, and besides it has an error - if `div` is an expression like `2 * 2` or `1 <<  3` it will result in incorrect computation, i suggest you remove it completely (or fix it if you don't want to remove it)

buha replied:
tbh, all of its uses in the file are okay, they are not expressions, they are either constants or variables, but someone's gonna come next and add a feature to the driver, use the macro with an expression and will waste effort debugging this

kister-jimenez replied:
Nice catch. Looking at the usage of this, I think it is shorter to just write the expression directly than use a macro for this. There's also not that many use.

buha commented on drivers/power/ltc4162l/iio_ltc4162l.c:
```diff
@@ -0,0 +1,514 @@
+		.name = _name,			\
+		.priv = _priv,			\
+		.show = ltc4162l_ch_attr_show,	\
+		.store = ltc4162l_ch_attr_store,	\
```
The newline characters are not aligned.

What i do is put a whitespace:
```
#define LTC4162L_IIO_CH_ATTR_RW(_name, _priv) \
	{ \
		.name = _name, \
		.priv = _priv, \
		.show = ltc4162l_ch_attr_show, \
		.store = ltc4162l_ch_attr_store, \
	}
```

There is nothing that can ever go wrong with this approach, no alignment to care about.

buha commented on drivers/power/ltc4162l/iio_ltc4162l.h:
```diff
@@ -0,0 +1,67 @@
+
+#include "iio.h"
+
+/******************************************************************************/
```
remove these, they are just noise:
```
/******************************************************************************/
/*************************** Types Declarations *******************************/
/******************************************************************************/
```

buha commented on projects/ltc4162l/src.mk:
```diff
@@ -0,0 +1,41 @@
+INCS += $(PROJECT)/src/platform/$(PLATFORM)/parameters.h
+SRCS += $(PROJECT)/src/platform/$(PLATFORM)/parameters.c
+
+INCS += $(INCLUDE)/no_os_delay.h     \
```
You can simplify this file by using NO_OS_INC_DIRS variable like:
```
NO_OS_INC_DIRS += $(INCLUDE)
```
... to replace all the list from line 14 to 27 for example

buha reviewed (commented):
this is a good PR but i think some changes still need to be made

@kister-jimenez yes, the xilinx build issue is unrelated, it was introduced by changes in analogdevicesinc/hdl repo which breaks no-OS CI builds every now and then...

mrcsosa commented:
Changelogs:

Removed:
* MULT_FRAC macro -> expression now directly used in the compute functions

Modified:
* src.mk now uses NO_OS_INC_DIRS to simplify include directories
* Code clean-up (iio_ltc4162l.c, iio_ltc4162l.h)
